runtime/gc_modron_standard/ReadBarrierVerifier.cpp

```diff
@@ -0,0 +1,292 @@
+		poisonJniWeakReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableMonitorObjectsReadBarrierVerification) {
+		poisonMonitorReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableClassStaticsReadBarrierVerification) {
+		poisonStaticClassSlots(env);
+		poisonConstantPoolObjects(env);
+	}
+}
+
+void
+MM_ReadBarrierVerifier::healSlot(MM_GCExtensionsBase *extensions, j9object_t *srcAddress)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t shadowHeapBase = (uintptr_t)extensions->shadowHeapBase;
+	uintptr_t shadowHeapTop = (uintptr_t)extensions->shadowHeapTop;
+	uintptr_t healedAddress;
+	j9object_t object = *srcAddress;
+
+	if (shadowHeapTop > (uintptr_t)object && (shadowHeapBase <= (uintptr_t)object)) {
+		healedAddress = heapBase + ((uintptr_t)object - shadowHeapBase);
+
+		// TODO: instead of atomicWriteReference, write back to srcAddress. But maybe we might need it to be atomic
+		*srcAddress = (j9object_t)healedAddress;
+	}
+}
+
+void
+MM_ReadBarrierVerifier::healSlot(MM_GCExtensionsBase *extensions, fomrobject_t *slot)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t shadowHeapBase = (uintptr_t)extensions->shadowHeapBase;
+	uintptr_t shadowHeapTop = (uintptr_t)extensions->shadowHeapTop;
+
+	uintptr_t healedHeapAddress;
+
+	uintptr_t referenceFromSlot = (uintptr_t)*slot;
+	if (referenceFromSlot < shadowHeapTop && (referenceFromSlot >= shadowHeapBase)) {
+
+		healedHeapAddress = heapBase + (referenceFromSlot - shadowHeapBase);
+		*slot = (fomrobject_t)healedHeapAddress;
+
+	}
+}
+
+void
+MM_ReadBarrierVerifier::healJniWeakReferenceSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true);
+	scanner.scanJNIWeakGlobalReferences(env);
+}
+
+void
+MM_ReadBarrierVerifier::healMonitorReferenceSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true);
+	scanner.scanMonitorReferences(env);
+}
+
+void
+MM_ReadBarrierVerifier::healStaticClassSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true);
+	scanner.scanClassStaticSlots(env);
+}
+
+void
+MM_ReadBarrierVerifier::healConstantPoolObjects(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true);
+	scanner.scanConstantPoolObjectSlots(env);
+}
+
+void
+MM_ReadBarrierVerifier::healSlots(MM_EnvironmentBase *env)
+{
+	MM_GCExtensionsBase *extensions = env->getExtensions();
+	if (1 == extensions->fvtest_enableJNIGlobalWeakReadBarrierVerification) {
+		healJniWeakReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableMonitorObjectsReadBarrierVerification) {
+		healMonitorReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableClassStaticsReadBarrierVerification) {
+		healStaticClassSlots(env);
+		healConstantPoolObjects(env);
+	}
+
+}
+
+bool
+MM_ReadBarrierVerifier::preObjectRead(J9VMThread *vmThread, J9Class *srcClass, j9object_t *srcAddress)
+{
```

Comment:
add the assert that vmThread provided is the calling thread (obtained from 'self thread' API in port library). add the same assert to the couple of other similar spots